File: apps/elixir/test/algorithm/strings_test.exs
defmodule Algorithm.StringsTest do
  use ExUnit.Case
  alias Algorithm.Strings

  describe "brute_force_search/2" do
    test "文字列内のパターンを検索する" do
      assert Strings.brute_force_search("ABCABC", "CAB") == {:ok, 2}
    end

    test "パターンが見つからない場合 :not_found を返す" do
      assert Strings.brute_force_search("ABCABC", "XYZ") == :not_found
    end
  end

  describe "kmp_search/2" do
    test "KMP 法で文字列を検索する" do
      assert Strings.kmp_search("ABCABC", "CAB") == {:ok, 2}
    end

    test "パターンが見つからない場合 :not_found を返す" do
      assert Strings.kmp_search("ABCABC", "XYZ") == :not_found
    end
  end

  describe "bm_search/2" do
    test "BM 法で文字列を検索する" do
      assert Strings.bm_search("ABCABC", "CAB") == {:ok, 2}
    end

    test "パターンが見つからない場合 :not_found を返す" do
      assert Strings.bm_search("ABCABC", "XYZ") == :not_found
    end

    test "先頭のパターンを見つける" do
      assert Strings.bm_search("ABCDEF", "ABC") == {:ok, 0}
    end
  end

  describe "char_count/1" do
    test "各文字の出現回数を返す" do
      assert Strings.char_count("hello") == %{"h" => 1, "e" => 1, "l" => 2, "o" => 1}
    end
  end

  describe "reverse_string/1" do
    test "文字列を逆順にする" do
      assert Strings.reverse_string("hello") == "olleh"
    end
  end

  describe "palindrome?/1" do
    test "回文の場合 true を返す" do
      assert Strings.palindrome?("racecar") == true
    end

    test "回文でない場合 false を返す" do
      assert Strings.palindrome?("hello") == false
    end
  end
end

File: apps/scala/src/test/scala/algorithm/StringsTest.scala
package algorithm

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class StringsTest extends AnyFunSuite with Matchers:

  test("bfMatch: パターンが見つかる"):
    Strings.bfMatch("ABCDEF", "CDE") shouldBe 2

  test("bfMatch: パターンが見つからない"):
    Strings.bfMatch("ABCDEF", "XYZ") shouldBe -1

  test("bfMatch: 空パターン"):
    Strings.bfMatch("ABCDEF", "") shouldBe 0

  test("kmpMatch: パターンが見つかる"):
    Strings.kmpMatch("AABCAABAACAABC", "AABC") shouldBe 0

  test("kmpMatch: パターンが見つからない"):
    Strings.kmpMatch("ABCDEF", "XYZ") shouldBe -1

  test("kmpMatch: 重複パターン"):
    Strings.kmpMatch("AABAACAADAABAABA", "AABA") shouldBe 0

  test("bmMatch: パターンが見つかる"):
    Strings.bmMatch("ABCDEF", "CDE") shouldBe 2

  test("bmMatch: パターンが見つからない"):
    Strings.bmMatch("ABCDEF", "XYZ") shouldBe -1

  test("countChars: 各文字の出現回数"):
    val result = Strings.countChars("aabbc")
    result('a') shouldBe 2
    result('b') shouldBe 2
    result('c') shouldBe 1

  test("reverseString: 文字列を逆順に"):
    Strings.reverseString("hello") shouldBe "olleh"
    Strings.reverseString("") shouldBe ""

  test("isPalindrome: 回文"):
    Strings.isPalindrome("racecar") shouldBe true
    Strings.isPalindrome("level") shouldBe true

  test("isPalindrome: 非回文"):
    Strings.isPalindrome("hello") shouldBe false

  test("isPalindrome: 空文字列"):
    Strings.isPalindrome("") shouldBe true
